Keen/Dataset/Datasets.cs

```diff
@@ -0,0 +1,301 @@
+
+            var serverBaseUrl = new Uri(prjSettings.KeenUrl);
+            _keenHttpClient = keenHttpClientProvider.GetForUrl(serverBaseUrl);
+            _cachedDatasetRelativeUrl = KeenHttpClient.GetRelativeUrl(prjSettings.ProjectId,
+                                                              KeenConstants.DatasetsResource);
+
+            _masterKey = prjSettings.MasterKey;
+        }
+
+        public async Task<JObject> Results(string datasetName, string indexBy, string timeframe)
+        {
+            if (string.IsNullOrWhiteSpace(datasetName))
+            {
+                throw new KeenException("A dataset name is required.");
+            }
+
+            if (string.IsNullOrWhiteSpace(indexBy))
+            {
+                throw new KeenException("A value to index by is required.");
+            }
+
+            if (string.IsNullOrWhiteSpace(timeframe))
+            {
+                throw new KeenException("A timeframe by is required.");
+            }
+
+            if (string.IsNullOrWhiteSpace(_masterKey))
+            {
+                throw new KeenException("An API masterkey is required to get dataset results.");
+            }
+
+            var datasetResultsUrl = $"{this.GetDatasetUrl(datasetName)}/results";
+
+            var url = $"{datasetResultsUrl}?index_by={indexBy}&timeframe={timeframe}";
+
+            var responseMessage = await _keenHttpClient
+                .GetAsync(url, _masterKey)
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            var responseString = await responseMessage
+                .Content
+                .ReadAsStringAsync()
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            var response = JObject.Parse(responseString);
+
+            KeenUtil.CheckApiErrorCode(response);
+
+            if (!responseMessage.IsSuccessStatusCode)
+            {
+                throw new KeenException("Request failed with status: " +
+                                        responseMessage.StatusCode);
+            }
+
+            return response;
+        }
+
+        public async Task<DatasetDefinition> Definition(string datasetName)
+        {
+            if (string.IsNullOrWhiteSpace(datasetName))
+            {
+                throw new KeenException("A dataset name is required.");
+            }
+
+            if (string.IsNullOrWhiteSpace(_masterKey))
+            {
+                throw new KeenException("An API masterkey is required to get dataset results.");
+            }
+
+            var responseMessage = await _keenHttpClient
+                .GetAsync(this.GetDatasetUrl(datasetName), _masterKey)
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            var responseString = await responseMessage
+                .Content
+                .ReadAsStringAsync()
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            var response = JObject.Parse(responseString);
```

Comment:
Here and elsewhere it might be nice to avoid parsing twice. Here we `JObject.Parse()` and then a few lines later unless there's an error condition, we `JsonConvert.DeserializeObject<>()`. I wonder if either by messing with `MissingMemberHandling` when deserializing, so that if it's an error object we still get the error info, or by using `JObject.ToObject<>()` after error checking we can do this without doing two full string parsing steps.

Reply:
See Issue #131